src/operator/tensor/indexing_op.cc

```diff
@@ -137,6 +137,26 @@ inline void SparseEmbeddingOpBackwardRspImpl<cpu>(const OpContext& ctx,
 }
 
 
+template<typename DType, typename IType>
+inline void ScatterNDAccForwardImpl(int N, int M, int K,
+                                    const mshadow::Shape<10> strides,
+                                    DType* out,
+                                    const DType* data,
+                                    const IType* indices,
+                                    mshadow::Stream<cpu> *s) {
+#pragma omp parallel for
+  for (int i = 0; i < N; i++) {
+    int offset = 0;
+    for (int j = 0; j < M; ++j) {
+      offset += strides[j] * static_cast<int>(indices[j*N + i]);
+    }
+    for (int j = 0; j < K; ++j) {
+#pragma omp atomic
+      out[offset + j] += data[i * K + j];
```

Comment:
You can consolidate this with the gpu kernel by using `#if __CUDA__ #else`in the header file since this line is the only difference between cpu and gpu kernels. Then in the `FCompute` function, you can use `Kernel::Launch` for both cpu and gpu kernels. That would make the implementation less verbose.

Reply:
@reminisce I've specialized the implementation of half_t and now it passes the test